File: ChromaSchemes/core/colour_generator.py
"""Deterministic colour generation from geometry expressions."""

from __future__ import annotations

from .chromaspace_adapter import ChromaSpaceAdapter


class ColourGenerator:
    """Generate RGB/OKLCH colours with stable IDs and origin metadata."""

    def __init__(self, adapter: ChromaSpaceAdapter | None = None) -> None:
        self._adapter = adapter or ChromaSpaceAdapter()

    def generate(
        self,
        base_hex: str,
        hue_offsets: list[float],
        chroma_values: list[float],
        luminance_values: list[float],
    ) -> list[dict]:
        base_rgb = self._adapter.hex_to_rgb(base_hex)
        base_h, _, _ = self._adapter.rgb_to_hsv(base_rgb)

        colours: list[dict] = []
        for index, offset in enumerate(hue_offsets):
            hue = (base_h + offset) % 360.0
            saturation = chroma_values[index]
            value = luminance_values[index]
            rgb = self._adapter.hsv_to_rgb(hue, saturation, value)
            oklch = self._adapter.hsv_to_oklch(hue, saturation, value)

            colours.append(
                {
                    "id": f"c{index}",
                    "origin": "base" if index < 4 else "derived",
                    "rgb": {"r": rgb[0], "g": rgb[1], "b": rgb[2]},
                    "oklch": {"l": oklch[0], "c": oklch[1], "h": oklch[2]},
                    "metadata": {"hueOffset": round(offset, 6)},
                }
            )

        return colours

File: ChromaSchemes/tooling/colourGenerator.ts
/**
 * ColourGenerator – converts geometry output into colour records.
 *
 * Description:
 *   Implements the colour-generation contract used by the ChromaSchemes
 *   tooling pipeline.
 *
 * Inputs:
 *   - generate(baseHex, hueOffsets, chromaValues, luminanceValues)
 *     * baseHex: base colour hex string.
 *     * hueOffsets: hue adjustments in degrees.
 *     * chromaValues: per-swatch chroma inputs.
 *     * luminanceValues: per-swatch luminance inputs.
 *
 * Outputs:
 *   An array of colour records containing RGB and OKLCh values.
 *
 * Example:
 *   const generator = new ColourGenerator();
 *   const colours = generator.generate('#ff8800', [0, 120], [0.2, 0.18], [0.8, 0.75]);
 */
import { IColourGenerator } from './interfaces';

export class ColourGenerator implements IColourGenerator {
  private hsvToRgb(h: number, s: number, v: number): { r: number; g: number; b: number } {
    const hue = ((h % 360) + 360) % 360;
    const sat = Math.max(0, Math.min(1, s));
    const val = Math.max(0, Math.min(1, v));
    const c = val * sat;
    const x = c * (1 - Math.abs(((hue / 60) % 2) - 1));
    const m = val - c;

    let rp = 0;
    let gp = 0;
    let bp = 0;
    if (hue < 60) [rp, gp, bp] = [c, x, 0];
    else if (hue < 120) [rp, gp, bp] = [x, c, 0];
    else if (hue < 180) [rp, gp, bp] = [0, c, x];
    else if (hue < 240) [rp, gp, bp] = [0, x, c];
    else if (hue < 300) [rp, gp, bp] = [x, 0, c];
    else [rp, gp, bp] = [c, 0, x];

    return {
      r: Math.round((rp + m) * 255),
      g: Math.round((gp + m) * 255),
      b: Math.round((bp + m) * 255),
    };
  }

  generate(baseHex: string, hueOffsets: number[], chromaValues: number[], luminanceValues: number[]) {
    const base = parseInt(baseHex.replace('#', ''), 16);
    const baseHue = ((base % 360) + 360) % 360;

    return hueOffsets.map((offset, idx) => {
      const hue = (baseHue + offset) % 360;
      const rgb = this.hsvToRgb(hue, chromaValues[idx], luminanceValues[idx]);
      return {
        id: `c${idx}`,
        origin: idx < 4 ? 'base' : 'derived',
        rgb,
        oklch: {
          l: Number(luminanceValues[idx].toFixed(6)),
          c: Number(chromaValues[idx].toFixed(6)),
          h: Number(hue.toFixed(6)),
        },
      };
    });
  }
}
